library/data_structures/seg_tree_uncommon/find_last.hpp

```diff
@@ -0,0 +1,34 @@
+/** @file */
+#pragma once
+/**
+ * @param le,ri defines range [le, ri)
+ * @param f defines a function that returns true if the subtree contains an
+ * element that satisfies the condition
+ * @returns the index of the last element in the range that satisfies the
+ * condition described in `f`, if no such element exists then (le - 1) is
+ * returned
+ * @time O(log(n))
+ * @space O(log(n)) for recursion stack; no new nodes are allocated
+ */
+template <class F>
+int find_last(int le, int ri, const F& f) {
+	return find_last_impl(le, ri, f, 0, n, 1);
+}
+template <class F>
+int find_last_impl(int le, int ri, const F& f, int tl, int tr, int u) {
+	if (ri <= tl || tr <= le) return le - 1;
+	if (tr - tl == 1) {
+		if (f(tree[u], tl, tr)) return tl;
+		return le - 1;
+	}
+	int tm = split(tl, tr);
+	push(tl, tm, tr, u);
+	if (le <= tl && tr <= ri) {
+		if (!f(tree[u], tl, tr)) return le - 1;
```

Comment:
why not return `ri`?

Reply:
cuz stl returns `end` iterator on not found

Reply:
Hmm it's possible that would be more consistent, but when working with things in "reverse" this still works which is nice

see: https://github.com/programming-team-code/programming_team_code/blob/6cebc5a0bcda539e3bb3e8e9580381018e623138/tests/library_checker_aizu_tests/handmade_tests/seg_tree_find.test.cpp#L36

Reply:
ah ok

Reply:
also comparing the index with some bound should still work if the code is written accordingly without an extra check for `ri` which can be annoying

ex. I always find myself checking `if (it == end(container))` which gets annoying